SI-6939 Fix namespace binding (xmlns) not overriding outer binding

eed3si9n opened the pull request:
This is a fix for [SI-6939: Namespace binding (xmlns) is duplicated if a child redefines a prefix](https://issues.scala-lang.org/browse/SI-6939).

Given a nested XML literal to the compiler Elem instance is generated
with namespace binding of the inner element copying that of the outer element:

``` scala
val foo = <x:foo xmlns:x="http://foo.com/">
  <x:bar xmlns:x="http://bar.com/">
    <x:baz/>
  </x:bar>
</x:foo>
```

With the above example, `foo.child.head.scope.toString` returns

``` scala
" xmlns:x="http://bar.com/" xmlns:x="http://foo.com/""
```

This is incorrect since the outer xmls:x should be overridden
by the inner binding.

XML library also parses XML document in a similar manner:

``` scala
val foo2 = scala.xml.XML.loadString("""<x:foo xmlns:x="http://foo.com/"><x:bar xmlns:x="http://bar.com/"><x:baz/></x:bar></x:foo>""")
```

Despite this erroneous behavior, since the structure of NamespaceBinding class
is designed to be singly-linked list, the stacking of namespace bindings allows
constant-time creation with simple implementation.
Since the inner namespace binding comes before the outer one,
query methods like `getURI` method behave correctly.

Because this bug is manifested when Elem is turned back into XML string,
it could be fixed by normalizing the namespace binding right when
buildString is called.
With this change `foo.child.head.scope.toString` now returns:

``` scala
" xmlns:x="http://bar.com/""
```

eed3si9n commented:
I asked on internal ML and @retronym said he'd review it. Thanks!

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2038/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2038/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1324/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1324/

retronym commented on src/library/scala/xml/NamespaceBinding.scala:
```diff
@@ -38,6 +38,22 @@ case class NamespaceBinding(prefix: String, uri: String, parent: NamespaceBindin
 
+  def normalize: NamespaceBinding = normalize(TopScope)
+
+  def normalize(stop: NamespaceBinding): NamespaceBinding = {
```
Could either or both of these be private? If not, can you make them final? (`class NameSpaceBinding` itself should be final like so many other classes, but we can't fix that without deprecation cycles).

eed3si9n replied:
I thought it might be useful to make those methods available publicly, but given the API promises Scala lib needs to maintain, maybe it's better to make them private.

retronym replied:
Yep, private is the way to go. We'll should then be able to target this at the 2.10.x branch (we've decided to not add new public methods in 2.10.N).

retronym commented on test/files/run/t6939.scala:
```diff
@@ -0,0 +1,7 @@
+  assert(foo.child.head.scope.toString == """ xmlns:x="http://bar.com/"""")
+
+  val foo2 = scala.xml.XML.loadString("""<x:foo xmlns:x="http://foo.com/"><x:bar xmlns:x="http://bar.com/"><x:baz/></x:bar></x:foo>""")
+  assert(foo2.child.head.scope.toString == """ xmlns:x="http://bar.com/"""")
```
Could you please add (at least) a test for this variation?

``` scala
scala> val foo = <foo xmlns="http://foo.com/"><bar xmlns="http://bar.com/"><baz/></bar></foo>
foo: scala.xml.Elem = <foo xmlns="http://foo.com/"><bar xmlns="http://bar.com/"><baz></baz></bar></foo>

scala> foo.child.head.scope.toString
res2: String = " xmlns="http://bar.com/" xmlns="http://foo.com/""
```

retronym commented:
In the bug report, you mention that:

> The same bug is duplicated in Scala library when XML document is created using scala.xml.XML.loadString(...) again due to outer namespace binding stacked in scala.xml.parsing.FactoryAdapter.

While it seems obvious enough that this fix will also address that, can you add a test to demonstrate?

eed3si9n commented:
@retronym wrote:

> While it seems obvious enough that this fix will also address that, can you add a test to demonstrate?

The `foo2` example uses `scala.xml.XML.loadString(...)`, which exercises the Scala library. With 2.10 you get both bar.com and foo.com:

``` scala
val foo2 = scala.xml.XML.loadString("""<x:foo xmlns:x="http://foo.com/"><x:bar xmlns:x="http://bar.com/"><x:baz/></x:bar></x:foo>""")
assert(foo2.child.head.scope.toString == """ xmlns:x="http://bar.com/"""")
```

retronym commented:
Oh sorry, that was right in front of me and I didn't see it.

eed3si9n commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2053/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2053/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1339/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1339/

eed3si9n commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1342/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2056/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1342/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2056/

eed3si9n commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2057/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2057/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1343/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1343/

paulp commented:
Sorry I'm late on review. Can you please:
- [ ] Rename 'normalize' to something more specific. We already have way too many methods called 'normalize', which sounds good but means nothing in isolation.
- [ ] Rebase and squash all this down to one commit.

Then it will LGTM.

eed3si9n committed:
SI-6939 Fix namespace binding (xmlns) not overriding outer binding

Given a nested XML literal to the compiler Elem instance is generated with
namespace binding of the inner element copying that of the outer element:

    val foo = <x:foo xmlns:x="http://foo.com/">
      <x:bar xmlns:x="http://bar.com/"><x:baz/></x:bar></x:foo>

With the above example, `foo.child.head.scope.toString` returns

    " xmlns:x="http://bar.com/" xmlns:x="http://foo.com/""

This is incorrect since the outer xmls:x should be overridden by the inner binding.

XML library also parses XML document in a similar manner:

    val foo2 = scala.xml.XML.loadString("""<x:foo xmlns:x="http://foo.com/"><x:bar xmlns:x="http://bar.com/"><x:baz/></x:bar></x:foo>""")

Despite this erroneous behavior, since the structure of NamespaceBinding class
is designed to be singly-linked list, the stacking of namespace bindings allows
constant-time creation with simple implementation.
Since the inner namespace binding comes before the outer one, query methods like `getURI` method behave correctly.

Because this bug is manifested when Elem is turned back into XML string, it could be fixed by shadowing the redefined namespace binding right when
buildString is called. With this change `foo.child.head.scope.toString` now returns:

    " xmlns:x="http://bar.com/""

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1622/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2353/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2353/ <br> ![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1622/ <br> ![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)

eed3si9n commented:
PLS REBUILD ALL

@paulp Thanks for the review. Renamed `normalize` to `shadowRedefined`, squashed to a single commit, and rebased it to the latest on master last night and ran the test.

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2368/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1622/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1622/ <br> ![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2368/ <br> ![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)

JamesIry commented:
One failure looks like a random Kitteh puke, the other looks legit. PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1709/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2444/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1711/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2446/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2444/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1711/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1709/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2446/

JamesIry commented:
@paulp, wanna give final stamp of approval? Seems all issues have been addressed.

paulp commented:
It's hard to believe that shadowRedefined method doesn't offer a simpler implementation, but I will not be staring it down. LGTM.
